T5577: Optimized PAM configs for RADIUS/TACACS+

zdc opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Optimized PAM configs for RADIUS/TACACS+

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://vyos.dev/T5577

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
groups, RADIUS, TACACS+, PAM

## Proposed changes
<!--- Describe your changes in detail -->
Now in CLI we can choose authentication logic for RADIUS and TACACS+ (current state is not changed and migration is not required):

  - `mandatory` - if an authentication server answered with `Access-Reject`/`REJECT`, authentication must be stopped and access denied immediately.
  - `optional` (default) - if an authentication server answers with `Access-Reject`/`REJECT`,  authentication continues using the next module.

In `mandatory` mode authentication will be stopped only if an authentication server clearly answered that access should be denied (no user in database, wrong password, etc.). If an authentication server is not available or other errors happen, it will be skipped and authentication will continue with the next module, like in `optional` mode.

This PR also fixes mistakes in PAM profiles (always ignored rules) and CLI speed issues when RADIUS/TACACS+ is configured.

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

zdc committed:
groups: T5577: Added `radius` and `tacacs` groups

We need separated groups for RADIUS and TACACS+ system users because they need
to be used in PAM rules independently.

zdc committed:
RADIUS: T5577: Added `mandatory` and `optional` modes for RADIUS

In CLI we can choose authentication logic:

  - `mandatory` - if RADIUS answered with `Access-Reject`, authentication must
  be stopped and access denied immediately.
  - `optional` (default) - if RADIUS answers with `Access-Reject`,
  authentication continues using the next module.

In `mandatory` mode authentication will be stopped only if RADIUS clearly
answered that access should be denied (no user in RADIUS database, wrong
password, etc.). If RADIUS is not available or other errors happen, it will be
skipped and authentication will continue with the next module, like in
`optional` mode.

zdc committed:
TACACS: T5577: Added `mandatory` and `optional` modes for TACACS+

In CLI we can choose authentication logic:

  - `mandatory` - if TACACS+ answered with `REJECT`, authentication must be
  stopped and access denied immediately.
  - `optional` (default) - if TACACS+ answers with `REJECT`, authentication
  continues using the next module.

In `mandatory` mode authentication will be stopped only if TACACS+ clearly
answered that access should be denied (no user in TACACS+ database, wrong
password, etc.). If TACACS+ is not available or other errors happen, it will be
skipped and authentication will continue with the next module, like in
`optional` mode.

c-po commented on debian/vyos-1x.postinst:
```diff
@@ -53,6 +48,11 @@ if grep -q '^tacacs' /etc/passwd; then
 
+# Remove TACACS+ PAM default profile
+if [[ -e /usr/share/pam-configs/tacplus ]]; then
+    rm /usr/share/pam-configs/tacplus
```
Deleting `/usr/share/pam-configs/tacplus` should not be required as the file is diverted in `debian/vyos-1x.preinst` script. We should not simply delete files but rather use debian tools and inform dpkg database about the changes.

zdc replied:
Unfortunately, this works only if a package installs something in place of a diverted file. In our case, there is no replacement and the file stays.

Also, diverted files will have another name, but `pam-auth-update` does not care about names and it will still use them.

c-po commented on interface-definitions/system-login.xml.in:
```diff
@@ -259,6 +259,25 @@
+                  </valueHelp>
+                  <constraint>
+                    <regex>(mandatory|optional)</regex>
+                  </constraint>
```
Can we make one of both a default setting to reflect the "old" behavior that was optional?

zdc replied:
First I wanted to say "It will require too much rewriting in `system-login.py`", but then found `from_defaults()`.

Done :)

zdc committed:
pam: T5577: Improved PAM configs for RADIUS and TACACS+

After sources analysis, we found the next possible return statuses for PAM
modules:

1. pam_tacplus

    Auth:

    - PAM_AUTH_ERR
    - PAM_AUTHINFO_UNAVAIL
    - PAM_AUTHTOK_ERR
    - PAM_BUF_ERR
    - PAM_CRED_INSUFFICIENT
    - PAM_PERM_DENIED
    - PAM_SUCCESS
    - PAM_USER_UNKNOWN

    Account:

    - PAM_AUTH_ERR
    - PAM_AUTHINFO_UNAVAIL
    - PAM_PERM_DENIED
    - PAM_SUCCESS
    - PAM_USER_UNKNOWN

    Session:

    - PAM_AUTHINFO_UNAVAIL
    - PAM_SESSION_ERR
    - PAM_SUCCESS
    - PAM_USER_UNKNOWN

2. pam_radius_auth

    Auth:

    - PAM_ABORT
    - PAM_AUTH_ERR
    - PAM_AUTHINFO_UNAVAIL
    - PAM_AUTHTOK_ERR
    - PAM_BAD_ITEM
    - PAM_BUF_ERR
    - PAM_CONV_AGAIN
    - PAM_CONV_ERR
    - PAM_IGNORE
    - PAM_NO_MODULE_DATA
    - PAM_PERM_DENIED
    - PAM_SUCCESS
    - PAM_SYSTEM_ERR
    - PAM_USER_UNKNOWN

    Account:

    - PAM_SUCCESS

    Session:

    - PAM_ABORT
    - PAM_AUTHINFO_UNAVAIL
    - PAM_BAD_ITEM
    - PAM_BUF_ERR
    - PAM_CONV_AGAIN
    - PAM_CONV_ERR
    - PAM_IGNORE
    - PAM_NO_MODULE_DATA
    - PAM_PERM_DENIED
    - PAM_SUCCESS
    - PAM_SYSTEM_ERR
    - PAM_USER_UNKNOWN

PAM configurations were replaced with tuned versions to take this into account.
